Add get-consul-client-ca command

ishustava opened the pull request:
When auto-encrypt is enabled, we need to retrieve Consul client CA from the Consul servers.

This command calls the `/agent/connect/ca/roots` endpoint, finds the currently active root CA, and writes it to the provided output file location.

This command allows you to provide a cloud-join string instead of the server address and this command will discover the servers. This allows us to re-use the `client.join` value in the Helm chart without requiring operators to provide the address of the server in addition to the join value.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,157 @@
+		"The server name to set as the SNI header when sending HTTPS requests to Consul. This can also be provided via the CONSUL_TLS_SERVER_NAME environment variable.")
+	c.flags.StringVar(&c.flagLogLevel, "log-level", "info",
+		"Log verbosity level. Supported values (in order of detail) are \"trace\", "+
+			"\"debug\", \"info\", \"warn\", and \"error\".")
```
I decided not to reuse consul's HTTP flags package here since we're not supposed to be using it.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,157 @@
+	c.flags.StringVar(&c.flagCAFile, "ca-file", "",
+		"The path to the CA file to use when making requests to the Consul server. This can also be provided via the CONSUL_CACERT environment variable")
+	c.flags.StringVar(&c.flagTLSServerName, "tls-server-name", "",
+		"The server name to set as the SNI header when sending HTTPS requests to Consul. This can also be provided via the CONSUL_TLS_SERVER_NAME environment variable.")
```
I decided not to reuse consul's HTTP flags package here since we're not supposed to be using it.

adilyse commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+
+	// wait for command to exit
+	retry.Run(t, func(r *retry.R) {
+		require.Equal(r, 0, exitCode)
```
This require was also included in the go routine for the command run. Are both needed? Is one a better spot than the other?

ishustava replied:
Nope, don't think we need it in both places.

adilyse commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+	// generate a new CA
+	ca, key := generateCA(t)
+
+	// set it as an active CA in Consul
```
It would be helpful if the comment mentioned that doing this causes the current CA to become inactive, making there be two roots.

adilyse commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+	require.NoError(err)
+
+	// Generate Server Cert
+	name := fmt.Sprintf("server.%s.%s", "dc1", "consul")
```
This seems like it's not necessary. Maybe just use `server.dc1.consul` as the `name` below?

lkysow commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		cfg.TLSConfig.Address = c.flagTLSServerName
+	}
+
+	return api.NewClient(cfg)
```
Note: this is what rebecca was saying above so wherever it makes most sense to document.

ishustava replied:
Yeah, it will use them, it was already documented in the command flags, and I've added Rebecca's suggestions for comments in the code. Good call!

lkysow commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		if err != nil {
+			return nil, err
+		}
+		logger.Debug("using cloud auto-join with", c.flagServerAddr)
```
For every arg after the first, you need to have key, value.
```suggestion
		logger.Debug("using cloud auto-join",  "server-addr", c.flagServerAddr)
```

Right now the logs look like:
```
2020-03-09T15:28:51.967-0400 [DEBUG] using cloud auto-join with: EXTRA_VALUE_AT_END=provider=fake
2020-03-09T15:28:51.967-0400 [DEBUG] discover: Using provider "fake"
2020-03-09T15:28:51.967-0400 [DEBUG] discovered servers: EXTRA_VALUE_AT_END=127.0.0.1
```

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+	// Wait until it gets a successful response
+	var activeRoot string
+	for activeRoot == "" {
+		caRoots, _, err := consulClient.Agent().ConnectCARoots(nil)
```
Yeah definitely, great idea! I'll keep the constant backoff for now, but curious if you think it's worth making it exponential.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+// newDiscover initializes the new Discover object
+// set up with all predefined providers, as well as
+// the k8s provider.
+func (c *Command) newDiscover() (*discover.Discover, error) {
```
Yep, will add a reference here and in the `consulClient` func

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		}
+	}
+
+	if c.flagCAFile != "" {
```
Good catch. I actually think it makes sense to not overwrite it since that is what [consul flags do](https://github.com/hashicorp/consul/blob/382d33bb7e21ea62d53089a26389ad375c297829/command/flags/http.go#L131-L150). I think it's good to keep this behavior consistent across, even though this is a separate binary.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/command/flags"
```
It's used for `c.help = flags.Usage(help, c.flags)`. I can move that over though.

ishustava replied:
On a second look, this might require a more comprehensive review of how to do that (This function looks a bit more complex than I hoped for). It's also used all over the repo and will definitely need to be addressed as part of not importing this library task. It probably doesn't make sense to spearhead it in this PR and feels a bit out of scope.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/command/flags"
+	"github.com/hashicorp/consul/lib"
```
Yeah, good point. I'll update it to say consul-k8s rather than Consul since that is the "product" that's making the go-discover call. That way we can also use the version of consul-k8s rather than Consul, which is easier to retrieve.

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+// write it to a file
+func TestRun(t *testing.T) {
+	t.Parallel()
+	outputFile, err := ioutil.TempFile("", "ca")
```
Great catch! I totally forgot to remove all those temp files 😱

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+	defer a.Stop()
+
+	// wait for command to exit
+	retry.Run(t, func(r *retry.R) {
```
It won't. By default it will timeout after 7 seconds
```
func Run(t Failer, f func(r *R)) {
	run(DefaultFailer(), t, f)
}
func DefaultFailer() *Timer {
	return &Timer{Timeout: 7 * time.Second, Wait: 25 * time.Millisecond}
}
```

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		// Pick the first server from the list,
+		// ignoring the port since we need to use HTTP API
+		firstServer := strings.SplitN(servers[0], ":", 2)[0]
+		cfg.Address = fmt.Sprintf("%s:8501", firstServer)
```
I like it! Will update!

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		// check if the server URL is missing a port
+		host := strings.TrimPrefix(c.flagServerAddr, "https://")
+		host = strings.TrimPrefix(c.flagServerAddr, "http://")
+		parts := strings.SplitN(host, ":", 2)
```
Yeah I tried using `url.Parse`. The problem with that function is that it's [ambiguous](https://golang.org/pkg/net/url/#Parse) if the scheme is missing. 

I don't think we should return error because my general intent with this was to assume secure defaults when not enough information about the server is provided (like in the case of cloud auto-join), but if someone really wants/needs to use HTTP, I think it should be allowed.

ishustava replied:
Also, it simplifies tests quite a bit because we don't have to generate certs for all test agents we're spinning up 😄

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+			return root.RootCertPEM, nil
+		}
+	}
+	return "", fmt.Errorf("none of the roots were active")
```
I like this because it doesn't require you to import the `errors` package, but it seems like a marginal difference to me. I will keep as-is for now.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+	for activeRoot == "" {
+		caRoots, _, err := consulClient.Agent().ConnectCARoots(nil)
+		if err != nil {
+			logger.Info("Error retrieving CA roots from Consul", "err", err)
```
Good catch!

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestRun_FlagsValidation(t *testing.T) {
```
- ca-file is not required, as I've mentioned above in case you're using HTTP
- server-addr will default to `127.0.0.1:8500`. Added this to the flag description
- added test for the invalid log level

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+// write it to a file
+func TestRun(t *testing.T) {
+	t.Parallel()
+	outputFile, err := ioutil.TempFile("", "ca")
```
Yes, the file will be created if it doesn't exist (under the hood, golang will make a system call that will create the file if it doesn't exist).

I went with the file because I like that you can simply write a string to it by calling `file.WriteString`. But I don't think it matters for the command. Curious why think dir is better.

lkysow replied:
I was thinking that you want to test that the command writes a file to disk. When the command runs normally the file it's writing to won't already exist but in this test you're pre-creating the file.

With a temporary dir you can test that it properly writes a file to that dir and then delete the temporary dir.

lkysow replied:
fwiw pre-creating a file when we want to test that it get's created doesn't make much sense to me. You're not calling `file.write` in this test.

ishustava replied:
To clarify, the intention here is to test that the expected contents are written to the file, not that the file gets created if it doesn't exist. The command should work the same in both cases. If the file already exists, it should write over the contents. If the file doesn't exist, it will create it and write to it. Testing those two cases exhaustively doesn't make much sense to me though because it's equivalent to testing golang's ioutil library. I think it's sufficient to trust the API's promise that it will create the file if it doesn't exist and just check that the file contents are what we expect.

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+	}()
+
+	// start the test server
+	time.Sleep(500 * time.Millisecond)
```
Yeah, good point, that is a bit too much. I've changed it to 100 ms. Although, it didn't really improve the test time too much since the slowest factor here is spinning up consul agents. Configuring the polling interval helps a bit more (from ~12sec to 11sec), but still doesn't feel significant. So my initial feeling is that adding a polling interval isn't worth it at the moment.

ishustava commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,227 @@
+		"Log verbosity level. Supported values (in order of detail) are \"trace\", "+
+			"\"debug\", \"info\", \"warn\", and \"error\".")
+
+	c.help = flags.Usage(help, c.flags)
```
That's interesting. It'll probably require rewriting the usage function, which I've commented about earlier. I also don't think it's worth fixing right now, but it'll hopefully auto-fix once we re-write that.

ishustava commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,388 @@
+			"-server-addr", fmt.Sprintf("http://127.0.0.1:%d", randomPorts[1]),
+			"-output-file", outputFile.Name(),
+		})
+		require.Equal(t, 0, exitCode)
```
yeah, I've updated it to instead run the server in the goroutine.

ishustava commented:
@adilyse @lkysow thanks so much for your reviews, they were super helpful 😄 

I've made some changes and refactors that came out of your suggestions that are ready for re-review. Summary here:
- To remove the dependency on the `tlsutil` package from Consul, I've refactored the `cert` package. I've pulled some generic cert generation functions out into its own file called `tls-util.go` and made functions in `source_gen.go` call them. This allowed us to re-use them for tests in this command and the server-acl-init command. I've made the changes to the server-acl-init tests in this PR, but happy to pull it out if you think it's better.
- After adding some tests for what was previously `consulClient` function's logic, I've found some mistakes in my implementation (thanks Rebecca for the suggestion!). I've changed the server address parsing to essentially better handle different possibilities for the flag value (different combinations of scheme and host or host and port). Generally, these are the cases:
   - If the scheme and port are provided, we don't do anything
   - If scheme without a port is provided, we look at the scheme and use the default port for that scheme
   - If the scheme is not provided, but only the host and port are provided, we use Consul's default scheme (HTTP).
   - If neither scheme or port are provided, we assume HTTPS port and scheme

lkysow commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,262 @@
+			return "", fmt.Errorf("could not discover any Consul servers with %q", c.flagServerAddr)
+		}
+
+		logger.Debug("discovered servers", strings.Join(servers, " "))
```
```suggestion
		logger.Debug("discovered servers", "servers", strings.Join(servers, " "))
```

lkysow commented on subcommand/get-consul-client-ca/command.go:
```diff
@@ -0,0 +1,262 @@
+	}
+
+	// Otherwise, return serverAddr:serverPort set by the provided flags
+	return fmt.Sprintf("%s:%s", c.flagServerAddr, c.flagServerPort), nil
```
I'd put this case first since it's shorter. i.e. `if !strings.contains(provider) { ... }`

lkysow commented on subcommand/get-consul-client-ca/command_test.go:
```diff
@@ -0,0 +1,428 @@
+	retry.Run(t, func(r *retry.R) {
+		require.NotNil(r, a)
+	})
+	defer a.Stop()
```
if `require.Equal(t, 0, exitCode)` fails then the server will never get stopped. Can you instead do
```
go func() { ... }
defer func() { if a != nil { a.Stop() }}
```

right after you spawn the goroutine?

ishustava replied:
Good catch! ty!

lkysow commented on subcommand/server-acl-init/command_test.go:
```diff
@@ -1397,34 +1395,21 @@ func generateServerCerts(t *testing.T) (string, string, string, func()) {
-
-	constraints := []string{"consul", "localhost"}
-	ca, err := tlsutil.GenerateCA(s, sn, 1, constraints)
+	signer, _, caCertPem, caCertTemplate, err := cert.GenerateCA("Consul Agent CA - Test")
```
:upvote: for having this in a helper function == minimal changes
